examples/pmatch.kl

```diff
@@ -0,0 +1,145 @@
+
+(meta
+  (datatype (Maybe A)
+    (nothing)
+    (just A)))
+
+(meta
+  (defun syntax->list (stx)
+    (syntax-case stx
+      [() (nil)]
+      [(cons x xs) (:: x (syntax->list xs))])))
+
+(meta
+  (define else-case
+     (lambda (stx)
+       (syntax-case stx
+         [(list (e x))
+          (if (identifier? e)
+              (free-identifier-case e
+                [else (pure (just x))]
+                [(else other) (pure (nothing))])
+              (pure (nothing)))]
+         [(ident e)
+          (free-identifier-case e
+            [else (pure (just 'else-x))]
+            [(else other) (pure (nothing))])]
+         [_ (pure (nothing))]))))
+
+(meta
+  (defun list->syntax (stx-lst stx)
+    (case stx-lst
+      [(nil) (empty-list-syntax stx)]
+      [(:: x xs) (cons-list-syntax x (list->syntax xs stx) stx)])))
+
+(meta
+  (datatype (Abstracted)
+    (none)
+    (bind Syntax Syntax (Abstracted)))
+
+  (defun abstract (stxs)
+    (syntax-case stxs
+      [() (none)]
+      [(cons e es)
+       (bind 'internal-x e (abstract es))])))
+   
+
+(define-macros
+  ([pmatch
+    (lambda (stx)
+      (syntax-case stx
+        [(cons _ more)
+         (syntax-case more
+           [(cons scrut pats)
+            (pure `(let (x ,scrut) (pmatch-aux ,scrut x ,pats)))])]))]
+   [pmatch-aux
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ scrut tgt pats))
+         (syntax-case pats
+           [() (pure `(error ',scrut))]
+           [(cons c cs)
+            (syntax-case c
+              [(list (pat rhs))
+               (>>= (else-case pat)
+                    (lambda (e)
+                      (case e
+                        [(just x) (pure `(let (,x ,tgt) ,rhs))]
+                        [(nothing)
+                         (pure `(let (fk (lambda (_) (pmatch-aux ,scrut ,tgt ,cs)))
+                                  (ppat ,tgt ,pat ,rhs (fk 'hi))))])))])])]))]
+   [ppat
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ tgt pat ks kf))
+         (syntax-case pat
+           [(ident x)
+            (pure `(let (,x ,tgt) ,ks))]
+           [(cons what args)
+            (>>= (make-temporaries (syntax->list args))
+                 (lambda (temps)
+                   (let (temp-names (map fst temps))
+                     (flet (combine (stxs)
+                             (case stxs
+                               [(nil) ks]
+                               [(:: id-and-stx rest)
+                                `(ppat ,(fst id-and-stx) ,(snd id-and-stx) ,(combine rest) ,kf)]))
+                        (pure `(case ,tgt
+                                 [,(cons-list-syntax what (the Syntax (list->syntax temp-names pat)) pat)
+                                  ,(combine temps)]
+                                 [(else other) ,kf]))))))])]))]))
+
+
+
+(datatype (Nat) (zero) (add1 (Nat)))
+
+(example (pmatch (zero) (else 't)))
+(example (pmatch (zero) ((else x) 't)))
+(example (pmatch (zero) (x 't)))
```

Comment:
`(else x)` and `x` both match anything? why do we need both?

Reply:
I bet I'm going to accidentally write the pattern `zero` instead of `(zero)` all the time. only having `,x` bind variables would be a lot less error-prone, I think (and would match the original algorithm)

Reply:
`else` unambiguously binds the variable `x`, while `x` alone may be a macro that expands to a pattern case using the normal `case` machinery.

As far as the comma goes, I am not a fan of the aesthetics of using an antiquotation operator for something that really isn't antiquotation...

Reply:
Just to be clear - `x` does not presently get to be a macro invocation, but this syntax at least does not rule that out!